lib/util/api.ts

```diff
@@ -14,190 +14,274 @@
+// const ETCHER_SERVER_ID = process.env.ETCHER_SERVER_ID as string;
 
-// NOTE: Ensure this isn't disabled, as it will cause
-// the stdout maxBuffer size to be exceeded when flashing
-ipc.config.silent = true;
+const ETCHER_TERMINATE_TIMEOUT: number = parseInt(
+	process.env.ETCHER_TERMINATE_TIMEOUT ?? '10000',
+	10,
+);
 
-// > If set to 0, the client will NOT try to reconnect.
-// See https://github.com/RIAEvangelist/node-ipc/
-//
-// The purpose behind this change is for this process
-// to emit a "disconnect" event as soon as the GUI
-// process is closed, so we can kill this process as well.
+const host = ETCHER_SERVER_ADDRESS ?? '127.0.0.1';
+const port = parseInt(ETCHER_SERVER_PORT || '3434', 10);
+// const path = ETCHER_SERVER_ID || "etcher";
 
-// @ts-ignore (0 is a valid value for stopRetrying and is not the same as false)
-ipc.config.stopRetrying = 0;
+// TODO: use the path as cheap authentication
 
-const DISCONNECT_DELAY = 100;
-const IPC_SERVER_ID = process.env.IPC_SERVER_ID as string;
-
-/**
- * @summary Send a message to the IPC server
- */
-function emit(channel: string, message?: any) {
-	ipc.of[IPC_SERVER_ID].emit(channel, message);
-}
-
-/**
- * @summary Send a log debug message to the IPC server
- */
-function log(message: string) {
-	if (console?.log) {
-		console.log(message);
-	}
-	emit('log', message);
-}
+const wss = new WebSocketServer({ host, port });
 
 /**
  * @summary Terminate the child process
  */
-async function terminate(exitCode: number) {
-	ipc.disconnect(IPC_SERVER_ID);
+async function terminate(exitCode?: number) {
 	await cleanup(Date.now());
 	process.nextTick(() => {
 		process.exit(exitCode || SUCCESS);
 	});
 }
 
-/**
- * @summary Handle errors
- */
-async function handleError(error: Error) {
-	emit('error', toJSON(error));
-	await delay(DISCONNECT_DELAY);
-	await terminate(GENERAL_ERROR);
-}
-
-/**
- * @summary Abort handler
- * @example
- */
-const onAbort = async (exitCode: number) => {
-	log('Abort');
-	emit('abort');
-	await delay(DISCONNECT_DELAY);
-	await terminate(exitCode);
-};
-
-const onSkip = async (exitCode: number) => {
-	log('Skip validation');
-	emit('skip');
-	await delay(DISCONNECT_DELAY);
-	await terminate(exitCode);
+// kill the process if no connections or heartbeat for  sec
+const setTerminateTimeout = () => {
+	if (ETCHER_TERMINATE_TIMEOUT > 0) {
+		return setTimeout(() => {
+			console.log(
+				`no connections or heartbeat for ${ETCHER_TERMINATE_TIMEOUT} ms, terminating`,
+			);
+			terminate();
+		}, ETCHER_TERMINATE_TIMEOUT);
+	} else {
+		return null;
+	}
 };
 
-ipc.connectTo(IPC_SERVER_ID, () => {
-	// Gracefully exit on the following cases. If the parent
-	// process detects that child exit successfully but
-	// no flashing information is available, then it will
-	// assume that the child died halfway through.
-
-	process.once('uncaughtException', handleError);
+let TerminateInterrupt = setTerminateTimeout();
+
+const setup = () =>
```

Comment:
Same, please convert this to a regular function. 

Also, it seems like only a few lines need to be within the promise constructor, please refactor it to limit its body.

Reply:
👍

Reply:
Actually, they all depends on the `connection` event, so apart from SIGINT/SIGTERM, everything else needs to be there. We could probably refactor differently to avoid that constraint but not in this cycle.